src/site/shared/circuitdesigner/impl/CircuitDesigner.ts

```diff
@@ -0,0 +1,67 @@
+import {Circuit, Obj}    from "core/public";
+import {DefaultTool}     from "shared/tools/DefaultTool";
+import {Tool}            from "shared/tools/Tool";
+import {Cursor}          from "shared/utils/input/Cursor";
+import {InputAdapter}    from "shared/utils/input/InputAdapter";
+import {CircuitDesigner} from "../CircuitDesigner";
+import {ToolManager}     from "./ToolManager";
+
+
+export class CircuitDesignerImpl<Circ extends Circuit> implements CircuitDesigner<Circ> {
+    public readonly circuit: Circ;
+
+    public inputAdapter: InputAdapter;
+
+    private readonly toolManager: ToolManager;
+
+    private readonly state: {
+        curPressedObj?: Obj;
+        cursor?: Cursor;
+    }
+
+    public constructor(
+        circuit: Circ,
+        { defaultTool, tools }: {
+            defaultTool: DefaultTool;
+            tools: Tool[];
+        },
+    ) {
+        this.circuit = circuit;
+
+        this.inputAdapter = new InputAdapter();
+
+        this.toolManager = new ToolManager(defaultTool, tools);
+
+        this.state = {
+            curPressedObj: undefined,
+            cursor:        undefined,
+        };
+
+        // Setup tool manager
+        this.inputAdapter.subscribe((ev) => this.toolManager.onEvent(ev, this));
+    }
+
+    public get curPressedObj() {
+        return this.state.curPressedObj;
+    }
+    public set curPressedObj(obj: Obj | undefined) {
+        this.state.curPressedObj = obj;
+    }
+
+    public get cursor() {
+        return this.state.cursor;
+    }
+    public set cursor(cursor: Cursor | undefined) {
+        this.state.cursor = cursor;
+    }
+
+    public attachCanvas(canvas: HTMLCanvasElement): () => void {
```

Comment:
It makes it so that in react with a `useEffect` we can `return designer.attachCanvas()` since the effects need to return the [cleanup function](https://react.dev/reference/react/useEffect#parameters) -- this is a **very** useful thing to have imo (and also will be described in the documentation eventually)

Reply:
I'm not really convinced eliding the cleanup function from the `useEffect` call this way is worth it, sacrificing clarity and separation of concern somewhat for lower syntactic overhead.  I'm willing to hold my judgement for now if you alias this return type so the semantics are clear, i.e. `ReactCleanupFunction`, emphasis on `React`.  The idea being to at some point revisit whether this is a pattern that makes sense to use at this level.

Reply:
I disagree, it's simply a "cleanup" function, feels like a not-super-crazy pattern ?
Otherwise, it would just return `void` anyways so might as well return something useful that the user doesn't necessarily have to use but can be convenient?